drivers/ethernet/eth_esp32.c

```diff
@@ -0,0 +1,290 @@
+		net_pkt_unref(pkt);
+		return NULL;
+	}
+
+	return pkt;
+}
+
+FUNC_NORETURN static void eth_esp32_rx_thread(void *arg1, void *arg2, void *arg3)
+{
+	const struct device *dev = arg1;
+	struct eth_esp32_dev_data *const dev_data = dev->data;
+
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	while (true) {
+		k_sem_take(&dev_data->int_sem, K_FOREVER);
+
+		uint32_t frames_remaining;
+
+		do {
+			struct net_pkt *pkt = eth_esp32_rx(
+				dev_data, &frames_remaining);
+			if (pkt == NULL) {
+				break;
+			}
+
+			if (net_recv_data(dev_data->iface, pkt) < 0) {
+				/* Upper layers are not ready to receive packets */
+				net_pkt_unref(pkt);
+			}
+		} while (frames_remaining > 0);
+	}
+}
+
+IRAM_ATTR static void eth_esp32_isr(void *arg)
+{
+	const struct device *dev = arg;
+	struct eth_esp32_dev_data *const dev_data = dev->data;
+	uint32_t intr_stat = emac_ll_get_intr_status(dev_data->hal.dma_regs);
+
+	emac_ll_clear_corresponding_intr(dev_data->hal.dma_regs, intr_stat);
+
+	if (intr_stat & EMAC_LL_DMA_RECEIVE_FINISH_INTR) {
+		k_sem_give(&dev_data->int_sem);
+	}
+}
+
+static int generate_mac(uint8_t mac_addr[6])
+{
+	int res = 0;
+#if DT_INST_PROP(0, zephyr_random_mac_address)
+	gen_random_mac(mac_addr, 0x24, 0xD7, 0xEB);
+#elif !NODE_HAS_VALID_MAC_ADDR(DT_DRV_INST(0))
+	if (esp_read_mac(mac_addr, ESP_MAC_ETH) != ESP_OK) {
+		res = -EIO;
+	}
+#endif
+	return res;
+}
+
+int eth_esp32_initialize(const struct device *dev)
+{
+	const struct eth_esp32_dev_config *const cfg = dev->config;
+	struct eth_esp32_dev_data *const dev_data = dev->data;
+	int res;
+
+	k_sem_init(&dev_data->int_sem, 1, 1);
+
+	periph_module_enable(PERIPH_EMAC_MODULE);
+
+	/* Convert 2D array DMA buffers to arrays of pointers */
+	for (int i = 0; i < CONFIG_ETH_DMA_RX_BUFFER_NUM; i++) {
+		dev_data->dma_rx_buf[i] = dev_data->dma->rx_buf[i];
+	}
+	for (int i = 0; i < CONFIG_ETH_DMA_TX_BUFFER_NUM; i++) {
+		dev_data->dma_tx_buf[i] = dev_data->dma->tx_buf[i];
+	}
+
+	emac_hal_init(&dev_data->hal, dev_data->dma->descriptors,
+		      dev_data->dma_rx_buf, dev_data->dma_tx_buf);
+
+	/* Configure ISR */
+	res = esp_intr_alloc(DT_IRQN(DT_NODELABEL(eth)),
+		       ESP_INTR_FLAG_IRAM,
+		       eth_esp32_isr,
+		       (void *)dev,
+		       NULL);
+	if (res != 0) {
+		goto err;
+	}
+
+	/* Configure phy for Media-Independent Interface (MII) or
+	 * Reduced Media-Independent Interface (RMII) mode
+	 */
+	if (strcmp(cfg->phy_connection_type, "rmii") == 0) {
+		emac_hal_iomux_init_rmii();
+		emac_hal_iomux_rmii_clk_input();
+		emac_ll_clock_enable_rmii_input(dev_data->hal.ext_regs);
+	} else if (strcmp(cfg->phy_connection_type, "mii") == 0) {
+		emac_hal_iomux_init_mii();
+		emac_ll_clock_enable_mii(dev_data->hal.ext_regs);
+	} else {
+		res = -EINVAL;
+		goto err;
+	}
+
+	/* Reset mac registers and wait until ready */
+	emac_ll_reset(dev_data->hal.dma_regs);
+	bool reset_success = false;
+
+	for (uint32_t t_ms = 0; t_ms < MAC_RESET_TIMEOUT_MS; t_ms += 10) {
+		k_sleep(K_MSEC(10));
```

Comment:
I'm wondering here, since the eth init runs at system initialization context if there is some risk of k_sleep fail here, it would not be better to use busy_wait instead?